### ISSUE TITLE
The adjacent matrix is very very sparse, how GCN works?

### ISSUE THREAD

**Issue by mensaochun** [chenzhaomin123/ML_GCN#6]
In the paper, the τ is set to 0.4 which lead to the best performance. But I find that if τ is set to 0.4, the adjacent matrix is very very sparse in voc data, that is to say, mainly class person connect to other classes. If many classes has not neighbors, how can gcn work? 
![image](https://user-images.githubusercontent.com/20737237/55797866-1a3f6280-5b00-11e9-8869-02d4e9abb638.png)

**Comment by DanZhang123** [comment]
hi@mensaochun, Do you know how to generate the adjacent matrix? In data directory，there exists voc_adj.pkl file,

**Comment by suanrong** [comment]
I use a diagonal matrix as adjacent matrix and get mAP of 92.5 on voc2007.

**Comment by WuXunjin** [comment]
> I use a diagonal matrix as adjacent matrix and get mAP of 92.5 on voc2007.

Hello can you please tell me how to generate the adjacent matrix?

**Comment by suanrong** [comment]
> > I use a diagonal matrix as adjacent matrix and get mAP of 92.5 on voc2007.
> 
> Hello can you please tell me how to generate the adjacent mat

if you mean the  voc_adj.pkl, just read the paper.
if you mean the _adj variables, just read the code in util.py

**Comment by suanrong** [comment]
After some test, I find max pooling play a key role for the performance improvement over the resnet101, which use avg pooling

**Comment by mensaochun** [comment]
> After some test, I find max pooling play a key role for the performance improvement over the resnet101, which use avg pooling

what is the performance difference?

**Comment by mensaochun** [comment]
> hi@mensaochun, Do you know how to generate the adjacent matrix? In data directory，there exists voc_adj.pkl file,
you need to compute the label concurrency.

**Comment by mensaochun** [comment]
> hi@mensaochun, Do you know how to generate the adjacent matrix? In data directory，there exists voc_adj.pkl file,

you need to compute the label concurency.

**Comment by Abhinav43** [comment]
@mensaochun  what do you mean by label concurency?

Suppose my labels looks like this : [ 1, 9 , 10 , 25 ]

How to generate adj matrix for this?

**Comment by Abhinav43** [comment]
@suanrong  How did you get that diagonal matrix?

I am confuses about voc_adj.pkl

**Comment by suanrong** [comment]
> > After some test, I find max pooling play a key role for the performance improvement over the resnet101, which use avg pooling
> 
> what is the performance difference?

The origin resnet101 get mAP of 89.5.
replacing the avg pooling of max pooling, we could get mAP of  92.5.

**Comment by Abhinav43** [comment]
@suanrong  can you please reply? How to get diagonal matrix?

If my labels looks like [ 1, 9 , 10 , 25 ] and diagonal matrix for this would be :

```
array([[ 1,  0,  0,  0],
       [ 0,  9,  0,  0],
       [ 0,  0, 10,  0],
       [ 0,  0,  0, 25]])

```

But some values are very big and some are very low, it will affect the GCN.

I was thinking one more thing so let's count the occurrence of labels something like : 

`occurrence = {   1 : 12 , 9 : 10  , 10: 19  , 25 : 26 } 
`
Now if we create  diagonal matrix of occurrence like : 

```
array([[12,  0,  0,  0],
       [ 0, 10,  0,  0],
       [ 0,  0, 19,  0],
       [ 0,  0,  0, 26]])
```
That would help?

Thank you !

**Comment by suanrong** [comment]
> > > > After some test, I find max pooling play a key role for the performance improvement over the resnet101, which use avg pooling
> > > 
> > > 
> > > what is the performance difference?
> > 
> > 
> > The origin resnet101 get mAP of 89.5.
> > replacing the avg pooling of max pooling, we could get mAP of 92.5.
> > Other metrics also get improvement of 1%~8%.
> 
> @suanrong , could you please share you hyper-parameters to train the network with GMP?
> I can not reach your accuracy(92.5%), my best result is around 91.0%.
> Thanks a lot!

python3 demo_voc2007_gcn.py data/voc --image-size 448 --batch-size 32 --epochs 40  --lr 0.001

**Comment by suanrong** [comment]
> @suanrong can you please reply? How to get diagonal matrix?
> 
> If my labels looks like [ 1, 9 , 10 , 25 ] and diagonal matrix for this would be :
> 
> ```
> array([[ 1,  0,  0,  0],
>        [ 0,  9,  0,  0],
>        [ 0,  0, 10,  0],
>        [ 0,  0,  0, 25]])
> ```
> But some values are very big and some are very low, it will affect the GCN.
> 
> I was thinking one more thing so let's count the occurrence of labels something like :
> 
> `occurrence = { 1 : 12 , 9 : 10 , 10: 19 , 25 : 26 } `
> Now if we create diagonal matrix of occurrence like :
> 
> ```
> array([[12,  0,  0,  0],
>        [ 0, 10,  0,  0],
>        [ 0,  0, 19,  0],
>        [ 0,  0,  0, 26]])
> ```
> That would help?
> 
> Thank you !

sorry ，I mean I use identity matrix as adj matrix here.

**Comment by Abhinav43** [comment]
@suanrong  If you mean like this :

```
array([[ 1.,  0.,  0.],
       [ 0.,  1.,  0.],
       [ 0.,  0.,  1.]])
```

Then there is no correlation between labels here, I mean it's totally random how it will help?

**Comment by suanrong** [comment]
> @suanrong If you mean like this :
> 
> ```
> array([[ 1.,  0.,  0.],
>        [ 0.,  1.,  0.],
>        [ 0.,  0.,  1.]])
> ```
> Then there is no correlation between labels here, I mean it's totally random how it will help?

I didn't say it would help

**Comment by Abhinav43** [comment]
@suanrong  But as you stated` "I use a diagonal matrix as adjacent matrix and get mAP of 92.5 on voc2007." `

That's why I was asking was it helpful?

**Comment by monk1337** [comment]
@suanrong  , @Abhinav43  , @DanZhang123 , @mensaochun , @kumamonatseu 

Can you guys take a look into this issue please 

https://github.com/chenzhaomin123/ML_GCN/issues/11

**Comment by zhanglonghao1992** [comment]
@mensaochun  After applying Eq. (8), the adjacent matrix A' should no longer be a sparse matrix.

**Comment by buptfanta** [comment]
> > > After some test, I find max pooling play a key role for the performance improvement over the resnet101, which use avg pooling
> > 
> > 
> > what is the performance difference?
> 
> The origin resnet101 get mAP of 89.5.
> replacing the avg pooling of max pooling, we could get mAP of 92.5.
> Other metrics also get improvement of 1%~8%.

Do it because the maxpooling is more sensitive to the attention area

**Comment by cgpeter96** [comment]
@suanrong 

Recently, I verified the validity of the GCN module . 
1. Replace the adj  by np.random.choice(0,2[80,80]),  the result as follow:
![image](https://user-images.githubusercontent.com/18066295/59925316-69900080-946a-11e9-8e24-a75f73e95e7f.png)

2. Replcae the label embedding by np.random.shuffle(inp), the result as follow:
![image](https://user-images.githubusercontent.com/18066295/59925366-89272900-946a-11e9-91af-0749286ad7a4.png)

and use resnet101（use maxpooling）can achieve comparable result （mAP 82.288）with ML_GCN(re-weight)>https://github.com/chenzhaomin123/ML_GCN/issues/13#issuecomment-504408587

emmmm,GCN module doesn't  work well !!!!!

**Comment by KMnP** [comment]
@cgpeter96 Hi, I wonder how did u generate the random adj? 
like this?
`np.random.choice([0, 2], [80, 80])`

**Comment by cgpeter96** [comment]
> @cgpeter96 Hi, I wonder how did u generate the random adj?
> like this?
> `np.random.choice([0, 2], [80, 80])`

= =i use `np.random.ranint(0,2,[80,80])`

**Comment by KMnP** [comment]
@cgpeter96  ahh.. I misread your previous post! Thank you for replying!
